### ISSUE TITLE
Focus outline issues

### ISSUE THREAD

**Issue by lyubomir-popov** [canonical-web-and-design/vanilla-framework#3058]
While QA-ing focus outlines, I noticed a few issues, I'll list them here:

![image](https://user-images.githubusercontent.com/2741678/81568639-f8b14a00-9395-11ea-8a00-1636e6eedfbb.png)

![image](https://user-images.githubusercontent.com/2741678/81573016-d8848980-939b-11ea-8dd1-732f64cdadcd.png)

**Comment by lyubomir-popov** [comment]
Awkward external link + outline combination:

![image](https://user-images.githubusercontent.com/2741678/81575053-7da06180-939e-11ea-9b92-3338febfece6.png)

**Comment by lyubomir-popov** [comment]
There's a bit of ambiguity when it comes to contrast between focus outline and inner background or border. Some WCAG [articles](https://www.w3.org/WAI/WCAG21/Understanding/non-text-contrast.html) talk about contrast with adjacent color, which is defined as the background outside of an element:

![image](https://user-images.githubusercontent.com/2741678/81692301-1c3bc980-9456-11ea-9a68-1ba16b0d792e.png)

![image](https://user-images.githubusercontent.com/2741678/81692458-560cd000-9456-11ea-87f6-41b390aeeaac.png)


But the focus outline goes on top of the border in the example of acceptable use:
![image](https://user-images.githubusercontent.com/2741678/81692593-8eaca980-9456-11ea-9c34-c4c98c24ac02.png)


Focus outlines need to have at least 3:1 contrast  with adjacent background:
![image](https://user-images.githubusercontent.com/2741678/81692728-bdc31b00-9456-11ea-9bc0-b0ef2322a2d0.png)


Meaning, if we put a neutral button on top of a gradient, we'd need to change the outline color to 3:1 contrast with the orange in this case:
![image](https://user-images.githubusercontent.com/2741678/81692807-db908000-9456-11ea-8720-3d61cdeccc3d.png)

There's no way to provide a focus outline that works against any background. This may mean we need a default focus for light backgrounds, and a modified one for dark backgrounds.

to compound things, there is no mention of contrast between button background (like the positive or negative buttons in vanilla) and the focus outline. To have sufficient contrast against both white and red, a near-black colour is the only option. But if you were to put that button in a dark strip, then you'd need a modifier to change the outline to a very light colour.

**Comment by lyubomir-popov** [comment]
According to [this](https://webaim.org/temp/1-4-11examples.htm) article, focus outlines should have 3:1 contrast with the default button state:



![image](https://user-images.githubusercontent.com/2741678/81694308-172c4980-9459-11ea-9296-cf4c0502b075.png)

**Comment by lyubomir-popov** [comment]
uk.gov article on focus states on their site:
https://design-system.service.gov.uk/get-started/focus-states/

**Comment by lyubomir-popov** [comment]
A very readable focus state that works on all backgrounds:

• >=3px border
• yellow that stands out on darker backgrounds where the dark border wouldn't be visible

![image](https://user-images.githubusercontent.com/2741678/81809026-8ddc4c00-9518-11ea-9836-14bc4462c17c.png)

**Comment by lyubomir-popov** [comment]
Applying focus on button click [varies](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/button) by OS/browser:

![image](https://user-images.githubusercontent.com/2741678/81925487-87ada480-95d8-11ea-8160-e3ad95de599d.png)

**Comment by lyubomir-popov** [comment]
In addition, :focus is not limited to outlines - if required on a button, it can change other properties like position, transform, background and still pas accessibility guidelines

**Comment by lyubomir-popov** [comment]
From this [article](https://css-tricks.com/almanac/selectors/f/focus-visible/), it seems :focus (and :focus-visible) is mandatory on elements that support user input. SO a button would be exempt.

**Comment by lyubomir-popov** [comment]
WCAG Success criterion: focus visible:
https://www.w3.org/WAI/WCAG21/Understanding/focus-visible
